### PR TITLE
Change Drawing Tests to use Perceptibility Threshold

### LINKED ISSUE #2107: Root-cause antialiasing differences in text rendering on build machines

### PR DESCRIPTION
**Pull request by aballway**
Differences between versions of Windows are generally imperceptibly small, so tests have been changed to ignore these differences.  Some tests have a few pixels where the differences are slightly beyond our threshold, so they still have the failure threshold as well.

Some test images have been updated to reflect updated output which fell behind slightly due to thresholds.  

Fixes #2107

### CODE REVIEW

**Comment by ms-jihua** [comment]
@DHowett-MSFT I neglected to mention this to you after you came back, but I have a proposal.

https://en.wikipedia.org/wiki/Color_difference I didn't read through all of it, but basically the 1976 primitive measurement of minimal perceptible color difference by the human eye is something like an RGBA difference of 2.3*256/100 = ~5.8.

Rather than shifting everything to mask, I think we should add a comparator that treats pixels as the same if they fall within an RGBA distance of... let's say half that, so 3. Thoughts?

**Comment by aballway** [comment]
Updated because they were failing on dev box with blue failures.  I think a few were still failing w/ perceptual comparison but don't want out of date images ykwim

File: tests/unittests/CoreGraphics.drawing/ImageComparison.cpp [hunk 1]
```diff
@@ -219,6 +219,30 @@ Pixel PixelComparisonModeMask<FailureThreshold>::ComparePixels(const LP& backgro
     return gp;
 }
 
+template <size_t FailureThreshold>
+template <typename LP, typename RP>
+Pixel PixelComparisonModeVisual<FailureThreshold>::ComparePixels(const LP& background, const LP& bp, const RP& cp, size_t& npxchg) {
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.cpp]: I feel that this name does not effectively convey what it actually checks for. 'Perceptible', 'HumanPerceptible', 'NoticeableDifference', or even 'Limen'/'DifferenceLimen' (if you want to be extra nerdy) would be more descriptive.

File: tests/unittests/CoreGraphics.drawing/ImageComparison.h [hunk 2]
```diff
@@ -24,11 +24,14 @@ enum class ImageComparisonResult : unsigned int { Unknown = 0, Incomparable, Dif
...
     uint8_t r, g, b, a;
     bool operator==(const Pixel& o) const {
-        return r == o.r && g == o.g && g == o.g && a == o.a;
+        return r == o.r && g == o.g && g == o.g && b == o.b && a == o.a;
     }
     bool operator!=(const Pixel& o) const {
         return !(*this == o);
     }
+    bool near(const Pixel& o) const {
+        return (abs(r - o.r) + abs(g - o.g) + abs(b - o.b) + abs(a - o.a)) < 4;
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.h]: should add a comment about that threshold

File: tests/unittests/CoreGraphics.drawing/ImageComparison.cpp (hunk 1 above)
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.cpp]: or 'CloseEnough' if you wanna be casual about it

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp [hunk 3]
```diff
@@ -131,7 +131,8 @@ static void __drawRightStackPink(CGContextRef context, CGRect bounds) {
     CGContextRestoreGState(context);
 }
 
-class CGContextBlendMode : public WhiteBackgroundTest<>, public ::testing::WithParamInterface<::testing::tuple<bool, CGNamedBlendMode>> {
+class CGContextBlendMode : public WhiteBackgroundTest<PixelByPixelImageComparator<PixelComparisonModeDifferenceLimen<100>>>,
```
Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: So blending still requires a threshold even with perceptual?

**Comment by DHowett-MSFT** [comment]
Nit: reword everything in title and body to better reflect what we are now doing

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp (hunk 3 above)
Comment (aballway) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: Mmhmm, there were between 1-75 pixels where the differences were beyond our threshold.

File: tests/unittests/CoreGraphics.drawing/ImageComparison.cpp [hunk 4]
```diff
@@ -171,53 +171,79 @@ struct RGBAImageBuffer {
...
+        gp.r = gp.g = gp.b = 0;
+        gp.a = 255;
+    }
+
+    return gp;
+}
+
+template <size_t FailureThreshold>
+template <typename LP, typename RP>
+Pixel PixelComparisonModeDifferenceLimen<FailureThreshold>::ComparePixels(const LP& background,
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.cpp]: oh darnit you actually went with it

File: tests/unittests/CoreGraphics.drawing/ImageComparison.h [hunk 5]
```diff
@@ -22,6 +22,8 @@
 enum class ImageComparisonResult : unsigned int { Unknown = 0, Incomparable, Different, Same };
 
 struct Pixel {
+    // Value at which pixel differences become perceptible
```
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.h]: I'm pretty sure that this isn't the actual value for just-noticeable-difference, just most likely under it. Could you change the comment to reflect that?

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp (hunk 3 above)
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: _how far_ beyond our threshold? related: do we want to be even more nice and change our difference images to reflect strength of difference between the two images? (possibly just change color coding to be abs(expected.r - actual.r) etc etc

File: tests/unittests/CoreGraphics.drawing/ImageComparison.cpp (hunk 4 above)
Comment (ms-jihua) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.cpp]: you totally should've used "EhCloseEnough" instead

File: tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp (hunk 3 above)
Comment (aballway) [inline tests/unittests/CoreGraphics.drawing/CGContextBlendModeTests.cpp]: It was something like 4-5 off.  I don't think we need to get that granular with our difference images.

File: tests/unittests/CoreGraphics.drawing/ImageComparison.cpp (hunk 4 above)
Comment (DHowett-MSFT) [inline tests/unittests/CoreGraphics.drawing/ImageComparison.cpp]: `PixelComparisonModeDifferenceIt_sFridayWhoCares`

**Pull request merged**